BCs (#134)
festim-dev/FESTIM-workshop: book/content/boundary_conditions/h_transport_basic.md

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.19.1
kernelspec:
  display_name: festim-workshop
  language: python
  name: python3
---

# Hydrogen transport: basic

This section discusses how to implement basic boundary conditions (BCs) for hydrogen transport problems. Boundary conditions are essential to FESTIM simulations, as they describe the mathematical problem at the boundaries of the simulated domain. Read more about BCs _[here](https://festim.readthedocs.io/en/fenicsx/userguide/boundary_conditions.html)_.

Objectives:
* Learn mathematical formulation for concentration and flux boundary conditions
* Implement fixed concentration boundary conditions
* Implement particle flux boundary conditions

+++

> ## Understanding math behind concentration and flux boundary conditions

Reviewer: Since this applies to all BCs (H transport and heat transfer) should this live in an introduction section?

### Fixed concentration

A fixed concentration (Dirichlet) boundary condition prescribes the value of the mobile hydrogen isotope concentration at a boundary. This enforces the concentration to remain constant in time and space on the specified boundary, independent of the solution in the bulk.

> This boundary condition is typically used to represent surfaces in equilibrium with an infinite reservoir, imposed implantation conditions, or experimentally controlled concentrations.

Reviewer: ```suggestion
This boundary condition is typically used to represent surfaces in equilibrium with a gas, imposed implantation conditions, or experimentally controlled concentrations.
```

#### Mathematical formulation

On a boundary $\Gamma_D$, the mobile concentration satisfies

$$
c(\mathbf{x}, t) = c_0 \quad \text{for } \mathbf{x} \in \Gamma_D,
$$

where $c_0$ is the prescribed concentration value.

> In the weak formulation, this condition is enforced by directly constraining the degrees of freedom associated with the boundary.

Reviewer: ```suggestion
This condition is enforced by directly constraining the degrees of freedom associated with the boundary.
```

DirichletBCs are not "weakly" enforced (ie not in the variational formulation)

+++

## Imposing fixed concentration boundary conditions

Users can prescribe a fixed concentration on a boundary using `FixedConcentrationBC`, which can depend on temperature, time, and space.

Let us consider a 2D, steady-state, single-species example with the following *space-dependent* boundary conditions:

$$ \text{Top:} \quad c = 2x + y $$
$$ \text{Right:} \quad c = y^2 + 1 $$

```{code-cell} ipython3
:tags: [hide-input]

import festim as F
import numpy as np
from dolfinx.mesh import create_unit_square
from mpi4py import MPI

my_model = F.HydrogenTransportProblem()
my_model.mesh = F.Mesh(create_unit_square(MPI.COMM_WORLD, 10, 10))
H = F.Species("H")
my_model.species = [H]
mat = F.Material(D_0=1e-3, E_D=0)
my_model.material = mat
my_model.temperature = 400
my_model.settings = F.Settings(atol=1e-8, rtol=1e-8, transient=False)
```

We can define the top and right boundary conditions using `lambda` functions:

```{code-cell} ipython3
top_bc_function = lambda x: 2.0 * x[0] + x[1]
right_bc_function = lambda x: x[1] ** 2 + 1.0

> ```

Reviewer: It may be of interest to add a `{note}` that a lambda function is just a shortcut for a `def func(x): ...` expression

Reviewer: I see you have this in the next section so maybe not required

```{note}

> `x[0]`, `x[1]`, `x[2]` corresponse to *(x, y, z)* in cartesian space. Additionally, boundary conditions can also be given as Python functions (see *{ref}`label-time-temp-concentration`* to learn more.)

Reviewer: ```suggestion
`x[0]`, `x[1]`, `x[2]` correspond to the first, second, and third space coordinage ( *(x, y, z)* in cartesian space). Additionally, boundary conditions can also be given as Python functions (see *{ref}`label-time-temp-concentration`* to learn more.)
```

> `x[0]`, `x[1]`, `x[2]` corresponse to *(x, y, z)* in cartesian space. Additionally, boundary conditions can also be given as Python functions (see *{ref}`label-time-temp-concentration`* to learn more.)

Reviewer: ```
Additionally, boundary conditions can also be given as Python functions
```
I don't get this, you already give it as a python function

```

+++

Now, let's create the boundaries to assign the BCs:

```{code-cell} ipython3
top_subdomain = F.SurfaceSubdomain(id=1, locator=lambda x: np.isclose(x[1], 1))
right_subdomain = F.SurfaceSubdomain(id=2, locator=lambda x: np.isclose(x[0], 1))
left_subdomain = F.SurfaceSubdomain(id=3, locator=lambda x: np.isclose(x[0], 0))
bottom_subdomain = F.SurfaceSubdomain(id=4, locator=lambda x: np.isclose(x[1], 0))
vol = F.VolumeSubdomain(id=5, material=mat)

> my_model.subdomains = [top_subdomain, right_subdomain, left_subdomain, bottom_subdomain, vol]

Reviewer: I would have this before defining the function, it breaks the flow a bit

D = F.Species("D")
boundary = F.SurfaceSubdomain(id=1)

top_H = F.FixedConcentrationBC(
    subdomain=boundary,
    value=5.0,
    species=H
)

top_D = F.FixedConcentrationBC(
    subdomain=boundary,
    value=10.0,
    species=D
)       
```

## Imposing particle flux boundary conditions

+++

Users can impose the particle flux at boundaries using `ParticleFluxBC` class. At minimum, this BC requires a boundary, value, and species to be added:

```{code-cell} ipython3
import festim as F

boundary = F.SurfaceSubdomain(id=1)
H = F.Species(name="H")

my_flux_bc = F.ParticleFluxBC(subdomain=boundary, value=2, species=H)
```

To use this BC to your problem, simply add it to `boundary_conditions` attribute of your problem:

```{code-cell} ipython3
my_model = F.HydrogenTransportProblem()
my_model.boundary_conditions = [my_flux_bc]
```

### Species-dependent fluxes

Similar to the concentration, the flux can be dependent on space, time and temperature. But for particle fluxes, the values can also be dependent on a species’ concentration. 

For example, let's define a hydrogen flux `J` that depends on the hydrogen concentration `c` and time `t`:

$$ J(c,t) = 10t^2 + 2c $$

> 

Reviewer: ```{note}
A flux boundary condition dependent on the actual solution is called a Robin boundary condition.
```
